Zend/tests/pipe_operator/call_by_ref.phpt

```diff
@@ -0,0 +1,20 @@
+--TEST--
+Pipe operator accepts by-reference functions
+--FILE--
+<?php
+
+function _modify(int &$a): string {
+    $a += 1;
+    return "foo";
+}
+
+//try {
+$a = 5;
+$res1 = $a |> '_modify';
```

Comment:
@Crell Is there any specific reason why we pipe a string? I'd like if we could do it without a string :)

```suggestion
$res1 = $a |> _modify;
```

Reply:
constants (potentially holding function names) vs bare function names - that won't work out.

Reply:
I think Crell did it in line with current callbacks definition we have in PHP:
```php
function foo(callable $f): void { $f(); }

foo('funcName'); // reference a function
foo([ $obj, 'methodName ]); // reference an instance method
foo([ ClassName::class, 'methodName' ]); // reference a static method
```

Reply:
Correct; as written it should accept any PHP-legal callable syntax.  PHP-legal callable syntax is kinda funky right now.  Improving the PHP-legal callable syntax is something else we should do, but is out of scope here.

Reply:
Why is it a callable and not a function call expression (as in Elixir, since you brought that up in your RFC)?

Taking your PSR7 example
```php
ServerRequest::fromGlobals()
    |> authenticate()
    |> [$router, 'resolveAction']()
    |> (fn($request) => $request->getAttribute('action')($request))()
// wouldn't people put this into a variable instead of defining inline?
// (as noted by nikic in the mailing list)
    |> renderResult()
    |> buildResponse()
    |> emit();
```

Reply:
Because function call expressions are a separate topic that should not be coupled to just pipe operators.  As the RFC notes, there's a separate RFC already written (but not coded, volunteers welcome) to handle partial function application, which gives us a nicer syntax in cases like this and any others, too.

Whether that RFC should block this one depends on who you ask. :smile:

Reply:
I don't see how callables are better, since they limit the usefulness of the pipe operator. You're limiting them to callables and a future partial function application.

With function call expressions, such as Elixir allows with the pipe operator, you'd actually get partial function applications instantly without having to worry about a future RFC, as the first parameter would be filled in (just as you'd write it out "inside out").

As the RFC currently stands, the pipe operator has only limited use, as it requires the given callable to only take a single parameter (or has default values for all following parameters). If we actually do a pipe operator, we should, in my opinion, do it right from the start and not limit it, in my opinion, arbitrarily to single parameter callables.

I'm by no means someone who has a lot to say in any PHP internals matters or even have the right to vote, but I do think I'm not the only one who thinks this way. I do not mean any offense if this comes off that way.

Reply:
I don't think anyone disagrees with you that callables are currently ugly in PHP; that's an existing PHP limitation.  However, a one-off syntax usable in only one place is not the answer.  Making callables easier to write in general is the answer.  The RFC links to another RFC that would resolve that, which would allow for:

```php
ServerRequest::fromGlobals()
    |> authenticate(?)
    |> $router->resolveAction(?)
    |> (fn($request) => $request->getAttribute('action')($request))()
    |> renderResult(?)
    |> buildResponse(?)
    |> emit(?);
```

Which would be ideal and cleaner than any other option currently available.

We just need to get people interested enough in that RFC to implement it...